llama_index/node_parser/unstructured_element.py

```diff
@@ -0,0 +1,303 @@
+        try:
+            response = query_engine.query(summary_query_str)
+            element.table_output = cast(PydanticResponse, response).response
+        except ValidationError as e:
+            # There was a pydantic validation error, so we will run with text completion
+            # fill in the summary and leave other fields blank
+            query_engine = index.as_query_engine()
+            response_txt = str(query_engine.query(summary_query_str))
+            element.table_output = TableOutput(summary=response_txt, columns=[])
+
+
+def get_table_elements(elements: List[Element]) -> List[Element]:
+    """Get table elements."""
+    return [e for e in elements if e.type == "table"]
+
+
+def get_text_elements(elements: List[Element]) -> List[Element]:
+    """Get text elements."""
+    return [e for e in elements if e.type == "text"]
+
+
+def _get_nodes_from_buffer(
+    buffer: List[str], node_parser: NodeParser
+) -> List[BaseNode]:
+    """Get nodes from buffer."""
+    doc = Document(text="\n\n".join(list(buffer)))
+    return node_parser.get_nodes_from_documents([doc])
+
+
+def get_nodes_from_elements(elements: List[Element]) -> List[BaseNode]:
+    """Get nodes and mappings."""
+    node_parser = SimpleNodeParser.from_defaults()
+
+    nodes = []
+    cur_text_el_buffer: List[str] = []
+    for element in elements:
+        if element.type == "table":
+            # flush text buffer
+            if len(cur_text_el_buffer) > 0:
+                cur_text_nodes = _get_nodes_from_buffer(cur_text_el_buffer, node_parser)
+                nodes.extend(cur_text_nodes)
+                cur_text_el_buffer = []
+
+            table_output = cast(TableOutput, element.table_output)
+            table_df = cast(pd.DataFrame, element.table)
+
+            table_id = element.id + "_table"
+            table_ref_id = element.id + "_table_ref"
+            # TODO: figure out what to do with columns
+            # NOTE: right now they're excluded from embedding
+            col_schema = "\n\n".join([str(col) for col in table_output.columns])
+            index_node = IndexNode(
+                text=str(table_output.summary),
+                metadata={"col_schema": col_schema},
+                excluded_embed_metadata_keys=["col_schema"],
+                id_=table_ref_id,
+                index_id=table_id,
+            )
+            table_str = table_df.to_string()
+            text_node = TextNode(
+                text=table_str,
+                id_=table_id,
+            )
+            nodes.extend([index_node, text_node])
+        else:
+            cur_text_el_buffer.append(str(element.element))
+
+    # flush text buffer
+    if len(cur_text_el_buffer) > 0:
+        cur_text_nodes = _get_nodes_from_buffer(cur_text_el_buffer, node_parser)
+        nodes.extend(cur_text_nodes)
+        cur_text_el_buffer = []
+
+    return nodes
+
+
+def get_base_nodes_and_mappings(nodes: List[BaseNode]) -> Tuple[List[BaseNode], Dict]:
```

Comment:
Any reason to make these functions outside of the class? Are they still used in the notebook?

Reply:
oh hm no particular reason. Maybe I'll move inside the class so users don't need to import two things